website/notifications/emails.py

```diff
@@ -0,0 +1,170 @@
+    notifications.get(notification_type)(subscribed_user_ids, uid, event, **context)
+
+
+def email_transactional(subscribed_user_ids, uid, event, **context):
+    """
+    :param subscribed_user_ids: mod-odm User object ids
+    :param uid: id of the event owner (Node or User)
+    :param event: name of notification event (e.g. 'comments')
+    :param context: context variables for email template
+    :return:
+    """
+    template = event + '.html.mako'
+    subject = Template(email_templates[event]['subject']).render(**context)
+
+    for user_id in subscribed_user_ids:
+        user = User.load(user_id)
+        email = user.username
+        context['localized_timestamp'] = localize_timestamp(context.get('timestamp'), user)
+        message = mails.render_message(template, **context)
+
+        if context.get('commenter')._id != user._id:
+            mails.send_mail(
+                to_addr=email,
+                mail=mails.TRANSACTIONAL,
+                mimetype='html',
+                name=user.fullname,
+                node_title=context.get('title'),
+                subject=subject,
+                message=message,
+                url=get_settings_url(uid, user)
+            )
+
+
+def email_digest(subscribed_user_ids, uid, event, **context):
+    """ Render the email message from context vars and store in the
+        DigestNotification objects created for each subscribed user.
+    """
+    template = event + '.html.mako'
+
+    try:
+        node = Node.find_one(Q('_id', 'eq', uid))
+        nodes = get_node_lineage(node, [])
+        nodes.reverse()
+    except NoResultsFound:
+        nodes = []
+
+    for user_id in subscribed_user_ids:
+        user = User.load(user_id)
+        context['localized_timestamp'] = localize_timestamp(context.get('timestamp'), user)
+        message = mails.render_message(template, **context)
+
+        if context.get('commenter')._id != user._id:
+            digest = DigestNotification(timestamp=context.get('timestamp'),
+                                        event=event,
+                                        user_id=user._id,
+                                        message=message,
+                                        node_lineage=nodes if nodes else [])
+            digest.save()
+
+
+def get_node_lineage(node, node_lineage):
+    """ Get a list of node ids in order from the parent project to the node itself
+        e.g. ['parent._id', 'node._id']
+    """
+    if node is not None:
+        node_lineage.append(node._id)
+    if node.node__parent != []:
+        for n in node.node__parent:
+            get_node_lineage(n, node_lineage)
+
+    return node_lineage
+
+
+def get_settings_url(uid, user):
+    if uid == user._id:
+        return web_url_for('user_notifications', _absolute=True)
+    else:
+        return Node.load(uid).absolute_url + 'settings/'
+
+
+def localize_timestamp(timestamp, user):
+    user_timezone = pytz.timezone(user.timezone)
```

Comment:
The chain of timezone usage relies on a 3rd party javascript extension. While it will probably work fine, there's nothing yet set to handle if it, for example, were to say that the timezone is set to `''`. If that happens, you will get an UnknownTimezoneError here. Probably catching it here and doing something sane would be the ideal thing to do (for example, not trying to localize the time but assuming GMT). Also, if it received a `None` here, you'd get an AttributeError. I don't know if it's possible to get a `None`, but an empty string or something equally invalid in string form seems very possible.